src/pool/srv_pool.c

```diff
@@ -5627,38 +5643,74 @@ pool_svc_schedule_reconf(struct pool_svc *svc)
 		return;
 	}
 
-	reconf_cancel_and_wait(reconf);
+	D_ASSERT(&svc->ps_reconf_sched == sched || &svc->ps_rfcheck_sched == sched);
+	sched_cancel_and_wait(sched);
 
-	/* Ended by pool_svc_reconf_ult. */
-	reconf_begin(reconf);
+	sched_begin(sched);
 
 	/*
 	 * An extra svc leader reference is not required, because
 	 * pool_svc_step_down_cb waits for this ULT to terminate.
 	 *
-	 * ULT tracking is achieved through svc->ps_reconf, not a ULT handle.
+	 * ULT tracking is achieved through sched, not a ULT handle.
 	 */
-	rc = dss_ult_create(pool_svc_reconf_ult, svc, DSS_XS_SELF, 0, 0, NULL /* ult */);
+	rc = dss_ult_create(func, svc, DSS_XS_SELF, 0, 0, NULL /* ult */);
 	if (rc != 0) {
-		D_ERROR(DF_UUID": failed to create reconfiguration ULT: "DF_RC"\n",
+		D_ERROR(DF_UUID": failed to create ULT: "DF_RC"\n",
 			DP_UUID(svc->ps_uuid), DP_RC(rc));
-		reconf_end(reconf);
+		sched_end(sched);
 	}
 
 	D_DEBUG(DB_MD, DF_UUID": end: "DF_RC"\n", DP_UUID(svc->ps_uuid), DP_RC(rc));
 }
 
-static void
-pool_svc_cancel_and_wait_reconf(struct pool_svc *svc)
-{
-	reconf_cancel_and_wait(&svc->ps_reconf);
-}
-
 static int pool_find_all_targets_by_addr(struct pool_map *map,
 					 struct pool_target_addr_list *list,
 					 struct pool_target_id_list *tgt_list,
 					 struct pool_target_addr_list *inval);
 
+static int
+cont_rf_check_cb(uuid_t pool_uuid, uuid_t cont_uuid, struct rdb_tx *tx, void *arg)
+{
+	struct pool_svc_sched *sched = arg;
+	int rc;
+
+	/* If anything happened during rf check, let's continue to check the next container
+	 * for the moment.
+	 */
+	rc = ds_cont_rf_check(pool_uuid, cont_uuid, tx);
+	if (rc)
+		D_CDEBUG(rc == -DER_RF, DB_MD, DLOG_ERR, DF_CONT" check_rf: "DF_RC"\n",
+			 DP_CONT(pool_uuid, cont_uuid), DP_RC(rc));
+
+	if (sched->psc_canceled) {
+		D_DEBUG(DB_MD, DF_CONT" is canceled.\n", DP_CONT(pool_uuid, cont_uuid));
+		return 1;
+	}
+
+	return 0;
+}
+
+static void
+pool_svc_rfcheck_ult(void *arg)
+{
+	struct pool_svc	*svc = arg;
+	int rc;
+
+	do {
+		/* retry until some one stop the pool svc(rc == 1) or succeed */
+		rc = ds_cont_rdb_iterate(svc->ps_uuid, cont_rf_check_cb,
+					 &svc->ps_rfcheck_sched);
+		if (rc >= 0)
+			break;
+		D_ERROR(DF_UUID" check rf with %d and retry\n", DP_UUID(svc->ps_uuid), rc);
+		dss_sleep(0);
```

Comment:
will this ULT be executed always even when there is no pool map update? that seems unnecessary, can be only be executed/wakeup when pool map changed?

Reply:
hmm, pool_svc_rfcheck_ult() is only scheduled for POOL_EXCLUDE and pool svc step up. and this while loop will break once it succeeds, (rc >= 0).